### ISSUE TITLE
documentation example and/or code modification for guild-specific member schemas

### ISSUE THREAD

**Issue by Lovinity** [dirigeants/klasa#474]
### Describe your proposal
I am proposing one or more of the following:
- Include an example on the Klasa documentation for creating a Schema in which one can store guild-specific data for members. Currently, there is documentation via the Users schema, but this is not guild-specific (the data applies to all guilds the user is in).
- Update the Schema system, if it is so necessary, such that there is a Members Schema by default in addition to Users; the Schema's data is separated by guild.

I do not know if it is possible to do guild-specific data Schemas with the current Klasa 0.5-dev. If it is possible, it is not documented well. If it is not possible, I propose it is made possible.

### Use-Cases for your proposal
- Points, XP, moderation logs, and other data that a developer may not want to be shared for a user across all guilds utilizing the same bot. For example, a developer may want users to have separate XP/level systems and rankings between each guild, rather than sharing the same XP/level across all guilds. A developer may especially not want moderation logs stored for members to be accessible between guilds as-is the case for the Users Schema.

### Expected and actual behavior
- This used to be achievable prior to #383 when Schemas were asynchronous (a schema key could be added along with a template for each new guild a user joins that uses the bot). However, that is no longer the case since this has been changed to synchronous.

### Further details
- This could potentially be achievable through the use of using objects within a Users schema, where the key is the guild ID, and the value is another object of data pertaining to that guild. However, by doing this, we miss out on Klasa verifying the data types used in the guild-specific objects. The end goal is to not only be able to have member data specific to each guild, but to enforce types on that data, just as we do in Schemas.

**Comment by KingDGrizzle** [comment]
You mean... The [Member Gateway Plugin?](https://github.com/dirigeants/klasa-member-gateway) That seems to provide the member schema you wish for, and it is super easy to install and use!

**Comment by Lovinity** [comment]
Not sure why I didn't find that when looking at the plugins lol. Disregard, and thank you!

**Comment by Tylertron1998** [comment]
Including documentation of how to make your own would be a little obsolete considering the plugin - and it's definitely possible to do on your own, too.

**Comment by Lovinity** [comment]
Understood. I didn't know the plugin existed until now. I do have one question... is it possible through Klasa to access / manipulate member-based data in the members gateway for a member who is no longer in the guild (eg. banning a member / adding a mod log for someone who already left)? Do I do it as I normally would for a member who is in the guild... or is there a different way / hack to doing that?

**Comment by Tylertron1998** [comment]
No, I don't think it would be possible. The settings instance sits on the GuildMember class, possibly with direct interaction with the provider, but for the use case you have, it would seem a little easier to store cases inside an object[] within the guilds settings.

**Comment by Lovinity** [comment]
There might be a way in which it can be done by accessing the settings directly via guildID.userID (which userID is the same as memberID). It wouldn't be possible to do a member.settings thing for a non-member, but I would imagine settings can still be accessed another way. I'll experiment, and if interested, will report my findings.

**Comment by Lovinity** [comment]
This will be my final comment on the issue unless I receive a reply.

@Tylertron1998 I have discovered a way in which one can access member records for people who are not in the guild. It's a little hacky but it works:
1. I extended Discord.js User to include a function, guildSettings(guildID), which returns this.client.gateways.members.create([guildID, this.id])
2. Instead of using member.settings, I use user.guildSettings(guildID).

Of course, this depends on the Member Gateway Plugin being installed.
